Mathlib/Topology/Compactness/Lindelof.lean

```diff
@@ -0,0 +1,709 @@
+  hf.toInducing.isLindelof_preimage (hf.closed_range) hK
+
+/-- A closed embedding is proper, ie, inverse images of Lindelöf sets are contained in Lindelöf.
+Moreover, the preimage of a Lindelöf set is Lindelöf, see `ClosedEmbedding.isLindelof_preimage`. -/
+theorem ClosedEmbedding.tendsto_coLindelof {f : X → Y} (hf : ClosedEmbedding f) :
+    Tendsto f (Filter.coLindelof X) (Filter.coLindelof Y) :=
+  hasBasis_coLindelof.tendsto_right_iff.mpr fun _K hK =>
+    (hf.isLindelof_preimage hK).compl_mem_coLindelof
+
+/-- Sets of subtype are Lindelöf iff the image under a coercion is. -/
+theorem Subtype.isLindelof_iff {p : X → Prop} {s : Set { x // p x }} :
+    IsLindelof s ↔ IsLindelof ((↑) '' s : Set X) :=
+  embedding_subtype_val.isLindelof_iff
+
+theorem isLindelof_iff_isLindelof_univ : IsLindelof s ↔ IsLindelof (univ : Set s) := by
+  rw [Subtype.isLindelof_iff, image_univ, Subtype.range_coe]
+
+theorem isLindelof_iff_LindelofSpace : IsLindelof s ↔ LindelofSpace s :=
+  isLindelof_iff_isLindelof_univ.trans isLindelof_univ_iff
+
+theorem IsLindelof.countable (hs : IsLindelof s) (hs' : DiscreteTopology s) : s.Countable :=
+  countable_coe_iff.mp
+  (@countable_of_Lindelof_of_discrete _ _ (isLindelof_iff_LindelofSpace.mp hs) hs')
+
+protected theorem ClosedEmbedding.nonLindelofSpace [NonLindelofSpace X] {f : X → Y}
+    (hf : ClosedEmbedding f) : NonLindelofSpace Y :=
+  nonLindelofSpace_of_neBot hf.tendsto_coLindelof.neBot
+
+protected theorem ClosedEmbedding.LindelofSpace [h : LindelofSpace Y] {f : X → Y}
+    (hf : ClosedEmbedding f) : LindelofSpace X :=
+  ⟨by rw [hf.toInducing.isLindelof_iff, image_univ]; exact hf.closed_range.isLindelof⟩
+
+/-- Countable topological spaces are Lindelof. -/
+instance (priority := 100) Countable.LindelofSpace [Countable X] : LindelofSpace X where
+  isLindelof_univ := countable_univ.isLindelof
+
+/-- The disjoint union of two Lindelöf spaces is Lindelöf. -/
+instance [LindelofSpace X] [LindelofSpace Y] : LindelofSpace (X ⊕ Y) :=
+  ⟨by
+    rw [← range_inl_union_range_inr]
+    exact (isLindelof_range continuous_inl).union (isLindelof_range continuous_inr)⟩
+
+instance {X : ι → Type*} [Countable ι] [∀ i, TopologicalSpace (X i)] [∀ i, LindelofSpace (X i)] :
+    LindelofSpace (Σi, X i) := by
+  refine' ⟨_⟩
+  rw [Sigma.univ]
+  exact isLindelof_iUnion fun i => isLindelof_range continuous_sigmaMk
+
+instance Quot.LindelofSpace {r : X → X → Prop} [LindelofSpace X] : LindelofSpace (Quot r) :=
+  ⟨by
+    rw [← range_quot_mk]
+    exact isLindelof_range continuous_quot_mk⟩
+
+instance Quotient.LindelofSpace {s : Setoid X} [LindelofSpace X] : LindelofSpace (Quotient s) :=
+  Quot.LindelofSpace
+
+/-- A set `s` is Hereditarily Lindelöf if every subset is a Lindelof set. We require this only
+for open sets in the definition, and then conclude that this holds for all sets by ADD. -/
+def IsHereditarilyLindelof (s : Set X) :=
+  ∀ t ⊆ s, IsOpen t → IsLindelof t
```

Comment:
I think that this definition doesn't work for any set `s`. E.g., if `s` has no interior points, then it is hereditarily Lindelöf in the sense of this definition. Should it be something like this?
```
∀ t, IsOpen t → IsLindelof (t ∩ s)
```
Please verify that your definition is correct by proving
```
lemma IsHereditarilyLindelof.isLindelof_subset (hs : IsHereditarilyLindelof s) (ht : t ⊆ s) :
    IsLindelof t := _

lemma IsHereditarilyLindelof.isLindelof (hs : IsHereditarilyLindelof s) :
    IsLindelof s := hs.isLindelof_subset Subset.rfl
```

Reply:
Perhaps requiring `IsOpen` is a bit clumsy, I'll just get rid of it here (as the official definition doesn't require it either!) and remove the lemma that concluded the more general case!